src/main/scala/edu/berkeley/ce/rockslicing/Block.scala

```diff
@@ -507,12 +513,23 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) {
       case ((volInc1, (centX1, centY1, centZ1)),  (volInc2, (centX2, centY2, centZ2))) =>
         (volInc1 + volInc2, (centX1 + centX2, centY1 + centY2, centZ1 + centZ2))
     }
-    Array(
-      // Factor of 3 comes from: centroid / (2.0 * (volume/6.0))
-      3.0 * centroidX / totalVolume,
-      3.0 * centroidY / totalVolume,
-      3.0 * centroidZ / totalVolume
-    )
+
+    // Check if block is extremely small
+    if (totalVolume <= NumericUtils.EPSILON) {
+      // If volume is essentially 0.0, return average of all vertices as centroid
+      val numberOfVertices = vertices.values.flatten.size
```

Comment:
Here's another way you could do this, which may or may not be better than what you have now.
```scala
val allVertices = vertices.values.flatten
Array[Double] (
  allVertices.map { v => v(0) }.sum / allVertices.length,
  allVertices.map { v => v(1) }.sum / allVertices.length,
  allVertices.map { v => v(2) }.sum / allVertices.length
)
```